lib/acyclicGraph.ml

```diff
@@ -581,50 +562,39 @@ module Make (Point:Point) = struct
      - we do a BFS rather than a DFS because we expect to have a short
          path (typically, the shortest path has length 1)
   *)
-  exception Found of canonical_node list
+  exception Found
   let search_path strict u v g =
-    let rec loop to_revert todo next_todo =
+    let rec loop status todo next_todo =
       match todo, next_todo with
-      | [], [] -> to_revert (* No path found *)
-      | [], _ -> loop to_revert next_todo []
+      | [], [] -> () (* No path found *)
+      | [], _ -> loop status next_todo []
       | (u, strict)::todo, _ ->
-        if u.status = Visited || (u.status = WeakVisited && strict)
-        then loop to_revert todo next_todo
-        else
-          let to_revert =
-            if u.status = NoMark then u::to_revert else to_revert
-          in
-          u.status <- if strict then WeakVisited else Visited;
+        if status_of status u = Visited || (status_of status u = WeakVisited && strict)
+        then loop status todo next_todo
+        else begin
```

Comment:
previous attempt was missing this begin causing a loop on
~~~coq
Inductive t : nat -> Set :=
|F1 : forall {n}, t (S n)
|FS : forall {n}, t n -> t (S n).

Definition case0 P (p: t 0): P p :=
  match p with | F1 | FS  _ => fun devil => False_rect (@IDProp) devil (* subterm !!! *) end.
~~~
(and others)